src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,346 @@
+  bdd_at_infty' := λ A, begin
+      rw slash_action.add_action,
+      exact (bounded_at_im_infty_subalgebra ℂ).add_mem' (f.bdd_at_infty' A) (g.bdd_at_infty' A),
+      end}}
+
+instance has_zero : has_zero (modular_form Γ k) :=
+{ zero := ⟨0, slash_action.mul_zero _, (λ _, mdifferentiable_at_const 𝓘(ℂ, ℂ) 𝓘(ℂ, ℂ)),
+  by {intro a,
+    convert (bounded_at_im_infty_subalgebra ℂ).zero_mem',
+    apply slash_action.mul_zero _ }⟩}
+
+instance has_nsmul : has_smul ℕ (modular_form Γ k) :=
+⟨ λ c f, {to_fun := (c : ℂ) • f,
+    slash_action_eq' := by {intro γ, convert slash_action.smul_action k γ f (c : ℂ),
+    exact ((f.slash_action_eq') γ).symm},
+    hol' := mdifferentiable.const_smul _ _ f.hol',
+    bdd_at_infty' := λ A, begin
+      rw slash_action.smul_action,
+      apply (bounded_at_im_infty_subalgebra ℂ).smul_mem (f.bdd_at_infty' A)
+    end }⟩
+
+instance has_zsmul : has_smul ℤ (modular_form Γ k) :=
+⟨ λ c f, {to_fun := (c : ℂ) • f,
+    slash_action_eq' := by {intro γ, convert slash_action.smul_action k γ f (c : ℂ),
+    exact ((f.slash_action_eq') γ).symm},
+    hol' := mdifferentiable.const_smul _ _ f.hol',
+    bdd_at_infty' := λ A, begin
+      rw slash_action.smul_action,
+      apply (bounded_at_im_infty_subalgebra ℂ).smul_mem (f.bdd_at_infty' A)
+    end }⟩
+
+instance has_csmul : has_smul ℂ (modular_form Γ k) :=
+⟨ λ c f, {to_fun := c • f,
+    slash_action_eq' := by {intro γ, convert slash_action.smul_action k γ f c,
+    exact ((f.slash_action_eq') γ).symm},
+    hol' := mdifferentiable.const_smul _ _ f.hol',
+    bdd_at_infty' := λ A, begin
+      rw slash_action.smul_action,
+      apply (bounded_at_im_infty_subalgebra ℂ).smul_mem (f.bdd_at_infty' A)
+    end }⟩
+
+instance has_neg : has_neg (modular_form Γ k) :=
+⟨λ f, {to_fun := -f,
+    slash_action_eq':= (-(f.to_slash_invariant_form)).2,
+    hol' := mdifferentiable.neg _ f.hol',
+    bdd_at_infty':= λ A, begin
+      convert (bounded_at_im_infty_subalgebra ℂ).smul_mem (f.bdd_at_infty' A) (-1),
+      simp only [_root_.neg_smul, one_smul],
+      apply  modular_forms.neg_slash,
+      end }⟩
+
+instance has_sub  : has_sub (modular_form Γ k) :=
+⟨λ f g, {to_fun := f - g,
+  slash_action_eq' :=
+    (slash_invariant_forms.has_sub.1 f.to_slash_invariant_form g.to_slash_invariant_form).2,
+    hol' := mdifferentiable.sub _ f.hol' g.hol',
+    bdd_at_infty' := λ A, begin
+      convert (bounded_at_im_infty_subalgebra ℂ).sub_mem (f.bdd_at_infty' A) (g.bdd_at_infty' A),
+      have :  (f : ℍ → ℂ) - g = f + (-g), by {funext, simp, ring,},
+      have h2 := slash_action.smul_action k A g.1 (-1 : ℂ),
+      simp only [_root_.neg_smul, one_smul, mf_to_fun_eq_coe] at h2,
+      rw [this, slash_action.add_action k A, h2],
+      refl,
+    end  }⟩
+
+instance : add_comm_group (modular_form Γ k) :=
+fun_like.coe_injective.add_comm_group _ rfl (λ _ _, by {refl}) (λ _, by{refl}) (λ _ _, by {refl})
+(λ _ _, by {simp, refl,}) (λ _ _, by {simp, refl})
```

Comment:
Each of these `λ _ _, by {refl}`s is a `coe_sub` etc lemma that you should write above